### ISSUE TITLE
io_uring deadlocks in the kernel during do_group_exit

### ISSUE THREAD

**Issue by dcolascione** [axboe/liburing#294]
I have a program that attempts to make several io_uring instances in different threads and poll them simultaneously. In some circumstances (which are hard to reproduce), if the program exits abruptly (e.g., because of a segfault or abort), something kernel-side hangs during cleanup and the task never exits. The task hang can't be cleared until the system reboots: SIGKILL does nothing. The CQs and SQs are likely full here.

The kernel is 5.10.13.

```
Feb 07 13:08:45 lug kernel: INFO: task fast_scan:1203716 blocked for more than 241 seconds.
Feb 07 13:08:45 lug kernel:       Tainted: G     U     OE     5.10.13-051013-generic #202102032337
Feb 07 13:08:45 lug kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Feb 07 13:08:45 lug kernel: task:fast_scan       state:D stack:    0 pid:1203716 ppid:     1 flags:0x00004000
Feb 07 13:08:45 lug kernel: Call Trace:
Feb 07 13:08:45 lug kernel:  __schedule+0x201/0x5d0
Feb 07 13:08:45 lug kernel:  schedule+0x4f/0xc0
Feb 07 13:08:45 lug kernel:  io_uring_cancel_files+0x13f/0x320
Feb 07 13:08:45 lug kernel:  ? wait_woken+0x80/0x80
Feb 07 13:08:45 lug kernel:  io_uring_cancel_task_requests+0x10c/0x240
Feb 07 13:08:45 lug kernel:  __io_uring_files_cancel+0x6f/0xc0
Feb 07 13:08:45 lug kernel:  do_exit+0xb9/0x3e0
Feb 07 13:08:45 lug kernel:  ? futex_wait+0x12c/0x240
Feb 07 13:08:45 lug kernel:  do_group_exit+0x3b/0xb0
Feb 07 13:08:45 lug kernel:  get_signal+0x165/0x600
Feb 07 13:08:45 lug kernel:  arch_do_signal+0x2f/0x1d0
Feb 07 13:08:45 lug kernel:  exit_to_user_mode_loop+0xab/0x130
Feb 07 13:08:45 lug kernel:  exit_to_user_mode_prepare+0x85/0x90
Feb 07 13:08:45 lug kernel:  syscall_exit_to_user_mode+0x2d/0x150
Feb 07 13:08:45 lug kernel:  do_syscall_64+0x45/0x90
Feb 07 13:08:45 lug kernel:  entry_SYSCALL_64_after_hwframe+0x44/0xa9
Feb 07 13:08:45 lug kernel: RIP: 0033:0x7f9ec78a3cd7
Feb 07 13:08:45 lug kernel: RSP: 002b:00007ffebac378d0 EFLAGS: 00000246 ORIG_RAX: 00000000000000ca
Feb 07 13:08:45 lug kernel: RAX: fffffffffffffe00 RBX: 00000000000000ca RCX: 00007f9ec78a3cd7
Feb 07 13:08:45 lug kernel: RDX: 0000000000125e1e RSI: 0000000000000000 RDI: 00007f9ec758d9d0
Feb 07 13:08:45 lug kernel: RBP: 0000000000000000 R08: 0000000000000000 R09: 00007f9ebdffb700
Feb 07 13:08:45 lug kernel: R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000125e1e
Feb 07 13:08:45 lug kernel: R13: 00007f9ec758d9d0 R14: 0000000000000000 R15: 00007f9ec758d700
```

**Comment by axboe** [comment]
Fixes for this just went into Linus's 5.11 tree, they should show up in 5.10-stable pretty soon.

**Comment by dcolascione** [comment]
Thanks
